### ISSUE TITLE
Release 4.2.0 - Release Candidate 12 - Workload benchmarks metrics

### ISSUE THREAD

**Issue by davidjiglesias** [wazuh/wazuh#9557]
The following issue aims to run all `workload benchmarks` for the current release candidate, report the results, and open new issues for any encountered errors.

## Workload benchmarks metrics information
|                                  |                                            |
|---------------------------------|--------------------------------------------|
| **Main release candidate issue** | https://github.com/wazuh/wazuh/issues/9457 |
| **Version** | 4.2.0                                      |
| **Release candidate #** | RC12                                         |
| **Tag** | https://github.com/wazuh/wazuh/tree/v4.2.0-rc12                                         |
| **Previous Workload benchmarks metrics issue** | --                                         |

## Test configuration
All tests will be run and workload performance metrics will be obtained for the following clustered environment configurations:
|                                  |                                            |
|---------------------------------|--------------------------------------------|
| **# Agents** | **# Worker nodes** |
| 1000 | 2 |
| 10000 | 10 |
| 50000 | 10 |
| 50000 | 25 |

## Test report procedure
All individual test checks must be marked as:
|                                  |                                            |
|---------------------------------|--------------------------------------------|
| Pass | The test ran successfully. |
| Xfail | The test was expected to fail and it failed. It must be properly justified and reported in an issue.  |
| Skip | The test was not run. It must be properly justified and reported in an issue.  |
| Fail | The test failed. A new issue must be opened to evaluate and address the problem. |

All test results must have one the following statuses: 
|                                  |                                            |
|---------------------------------|--------------------------------------------|
| :green_circle:  | All checks passed. |
| :red_circle:  | There is at least one failed check. |
| :yellow_circle:  | There is at least one expected fail or skipped test and no failures. |

Any failing test must be properly addressed with a new issue, detailing the error and the possible cause. It must be included in the `Fixes` section of the current release candidate main issue https://github.com/wazuh/wazuh/issues/9457.

Any expected fail or skipped test must have an issue justifying the reason. All auditors must validate the justification for an expected fail or skipped test.

An extended report of the test results must be attached as a zip or txt. This report can be used by the auditors to dig deeper into any possible failures and details.

## Conclusions

All tests have been executed and the results can be found [here](https://github.com/wazuh/wazuh/issues/9557#issuecomment-895141039).

All tests have passed and the fails have been reported or justified. I therefore conclude that this issue is finished and OK for this release candidate.

For a detailed conclusion and report on the cluster performance metrics please refer to [1000/2 10000/10](https://github.com/wazuh/wazuh/issues/9557#issuecomment-896001756) and [50000/10 50000/25](https://github.com/wazuh/wazuh/issues/9557#issuecomment-896535717).

## Auditors validation
The definition of done for this one is the validation of the conclusions and the test results from all auditors.

All checks from below must be accepted in order to close this issue.

- [ ] @santiago-bassett 
- [x] @davidjiglesias 
- [x] @AdriiiPRodri  
- [x] @vicferpoy

**Comment by mcarmona99** [comment]
## Extended Test Report

report.zip

## Test Results

### Run 1000 agents - 2 workers

---
|                                  |                                            |                                           |            
|--|--|--|
| **Endpoint test name**  | **Status** | **Issues Ref.** |
| GET /cluster/local/info | :green_circle: | |
| GET /cluster/nodes| :green_circle:  | |
| GET /cluster/healthcheck | :green_circle: | |
| GET /cluster/status | :green_circle: | |
| GET /cluster/local/config | :green_circle: | |
| GET /cluster/api/config | :green_circle: | |
| GET /cluster/configuration/validation | :green_circle: | |
| GET /manager/status | :green_circle: | |
| GET /manager/info | :green_circle: | |
| GET /manager/configuration | :green_circle: | |
| GET /manager/logs | :green_circle: | |
| GET /manager/api/config | :green_circle: | |
| GET /manager/configuration/validation | :green_circle: | |
| GET /mitre | :green_circle: | |
| GET /overview/agents | :green_circle:  | |
| GET /tasks/status | :green_circle: | |
| PUT /active-response | :yellow_circle: | https://github.com/wazuh/wazuh-qa/issues/1266 |
| PUT /rootcheck | :green_circle: | |
| PUT /syscheck | :green_circle: | |
| DELETE /rootcheck | :green_circle: | |
| POST /groups | :green_circle: | |
| POST /security/users | :green_circle: | |
| POST /security/roles | :green_circle: | |
| POST /security/policies | :green_circle: | |
| POST /security/rules | :green_circle: | |
| PUT /agents/group | :green_circle: | |
| PUT /agents/group/new_test_group/restart | :green_circle: | |
| PUT /agents/restart | :green_circle: | |
| POST /agents | :green_circle: | |
| POST /agents/insert​|  :green_circle: | |
| POST /agents/insert​/quick | :green_circle: | |
| GET /agents | :green_circle: | |
| GET /agents/no_group | :green_circle: | |
| GET /agents/outdated | :green_circle: | |
| GET /decoders | :green_circle: | |
| GET /groups | :green_circle: | |
| GET /lists | :green_circle: | |
| GET /rules | :green_circle: | |
| GET /security/users | :green_circle: | |
| GET /security/roles | :green_circle: | |
| GET /security/policies | :green_circle: | |
| GET /security/rules | :green_circle: | |
| DELETE /agents | :green_circle: | |
| DELETE /groups | :green_circle: | |
| DELETE /security/users | :green_circle: | |
| DELETE /security/roles | :green_circle: | |
| DELETE /security/policies | :green_circle: | |
| DELETE /security/rules | :green_circle: | |
| PUT /manager/restart | :green_circle: | |
| PUT /cluster/restart | :green_circle: | |

### Run 10000 agents - 10 workers

---
|                                  |                                            |                                           |            
|--|--|--|
| **Endpoint test name**  | **Status** | **Issues Ref.** |
| GET /cluster/local/info | :green_circle: | |
| GET /cluster/nodes| :green_circle:  | |
| GET /cluster/healthcheck | :green_circle: | |
| GET /cluster/status | :green_circle: | |
| GET /cluster/local/config | :green_circle: | |
| GET /cluster/api/config | :green_circle: | |
| GET /cluster/configuration/validation | :green_circle: | |
| GET /manager/status | :green_circle: | |
| GET /manager/info | :green_circle: | |
| GET /manager/configuration | :green_circle: | |
| GET /manager/logs | :green_circle: | |
| GET /manager/api/config | :green_circle: | |
| GET /manager/configuration/validation | :green_circle: | |
| GET /mitre | :green_circle: | |
| GET /overview/agents | :green_circle:  | |
| GET /tasks/status | :green_circle: | |
| PUT /active-response | :yellow_circle: | https://github.com/wazuh/wazuh-qa/issues/1266 |
| PUT /rootcheck | :green_circle: | |
| PUT /syscheck | :green_circle: | |
| DELETE /rootcheck | :red_circle: | #8852 (Fixed in `4.3.0`) |
| POST /groups | :green_circle: | |
| POST /security/users | :green_circle: | |
| POST /security/roles | :green_circle: | |
| POST /security/policies | :green_circle: | |
| POST /security/rules | :green_circle: | |
| PUT /agents/group | :green_circle: | |
| PUT /agents/group/new_test_group/restart | :green_circle: | |
| PUT /agents/restart | :green_circle: | |
| POST /agents | :green_circle: | |
| POST /agents/insert​|  :green_circle: | |
| POST /agents/insert​/quick | :green_circle: | |
| GET /agents | :green_circle: | |
| GET /agents/no_group | :green_circle: | |
| GET /agents/outdated | :green_circle: | |
| GET /decoders | :green_circle: | |
| GET /groups | :green_circle: | |
| GET /lists | :green_circle: | |
| GET /rules | :green_circle: | |
| GET /security/users | :green_circle: | |
| GET /security/roles | :green_circle: | |
| GET /security/policies | :green_circle: | |
| GET /security/rules | :green_circle: | |
| DELETE /agents | :red_circle: | #8853 (Fixed in `4.3.0`) |
| DELETE /groups | :red_circle: | #9014 (Fixed in `4.3.0`) |
| DELETE /security/users | :green_circle: | |
| DELETE /security/roles | :green_circle: | |
| DELETE /security/policies | :green_circle: | |
| DELETE /security/rules | :green_circle: | |
| PUT /manager/restart | :green_circle: | |
| PUT /cluster/restart | :green_circle: | |

### Run 50000 agents - 10 workers

---
|                                  |                                            |                                           |            
|--|--|--|
| **Endpoint test name**  | **Status** | **Issues Ref.** |
| GET /cluster/local/info | :green_circle: | |
| GET /cluster/nodes| :green_circle:  | |
| GET /cluster/healthcheck | :green_circle: | |
| GET /cluster/status | :green_circle: | |
| GET /cluster/local/config | :green_circle: | |
| GET /cluster/api/config | :green_circle: | |
| GET /cluster/configuration/validation | :green_circle: | |
| GET /manager/status | :green_circle: | |
| GET /manager/info | :green_circle: | |
| GET /manager/configuration | :green_circle: | |
| GET /manager/logs | :green_circle: | |
| GET /manager/api/config | :green_circle: | |
| GET /manager/configuration/validation | :green_circle: | |
| GET /mitre | :green_circle: | |
| GET /overview/agents | :green_circle:  | |
| GET /tasks/status | :green_circle: | |
| PUT /active-response | :yellow_circle: | https://github.com/wazuh/wazuh-qa/issues/1266 |
| PUT /rootcheck | :red_circle: | #8865 (Fixed in `4.3.0`) |
| PUT /syscheck | :red_circle: | #8867 (Fixed in `4.3.0`) |
| DELETE /rootcheck | :red_circle: | #8852 (Fixed in `4.3.0`) |
| POST /groups | :green_circle: | |
| POST /security/users | :green_circle: | |
| POST /security/roles | :green_circle: | |
| POST /security/policies | :green_circle: | |
| POST /security/rules | :green_circle: | |
| PUT /agents/group | :red_circle: | #6749 (Fixed in `4.3.0`) |
| PUT /agents/group/new_test_group/restart | :red_circle: | #8870 (Fixed in `4.3.0`) |
| PUT /agents/restart | :red_circle: | #8870 (Fixed in `4.3.0`) |
| POST /agents | :green_circle: | |
| POST /agents/insert​|  :green_circle: | |
| POST /agents/insert​/quick | :green_circle: | |
| GET /agents | :green_circle: | |
| GET /agents/no_group | :green_circle: | |
| GET /agents/outdated | :green_circle: | |
| GET /decoders | :green_circle: | |
| GET /groups | :green_circle: | |
| GET /lists | :green_circle: | |
| GET /rules | :green_circle: | |
| GET /security/users | :green_circle: | |
| GET /security/roles | :green_circle: | |
| GET /security/policies | :green_circle: | |
| GET /security/rules | :green_circle: | |
| DELETE /agents | :red_circle: | #8853 (Fixed in `4.3.0`) |
| DELETE /groups | :red_circle: | #9014 (Fixed in `4.3.0`) |
| DELETE /security/users | :green_circle: | |
| DELETE /security/roles | :green_circle: | |
| DELETE /security/policies | :green_circle: | |
| DELETE /security/rules | :green_circle: | |
| PUT /manager/restart | :green_circle: | |
| PUT /cluster/restart | :green_circle: | |

### Run 50000 agents - 25 workers

---
|                                  |                                            |                                           |            
|--|--|--|
| **Endpoint test name**  | **Status** | **Issues Ref.** |
| GET /cluster/local/info | :green_circle: | |
| GET /cluster/nodes| :green_circle:  | |
| GET /cluster/healthcheck | :green_circle: | |
| GET /cluster/status | :green_circle: | |
| GET /cluster/local/config | :green_circle: | |
| GET /cluster/api/config | :green_circle: | |
| GET /cluster/configuration/validation | :green_circle: | |
| GET /manager/status | :green_circle: | |
| GET /manager/info | :green_circle: | |
| GET /manager/configuration | :green_circle: | |
| GET /manager/logs | :green_circle: | |
| GET /manager/api/config | :green_circle: | |
| GET /manager/configuration/validation | :green_circle: | |
| GET /mitre | :green_circle: | |
| GET /overview/agents | :green_circle:  | |
| GET /tasks/status | :green_circle: | |
| PUT /active-response | :yellow_circle: | https://github.com/wazuh/wazuh-qa/issues/1266 |
| PUT /rootcheck | :red_circle: | #8865 (Fixed in `4.3.0`) |
| PUT /syscheck | :red_circle: | #8867 (Fixed in `4.3.0`) |
| DELETE /rootcheck | :red_circle: | #8852 (Fixed in `4.3.0`) |
| POST /groups | :green_circle: | |
| POST /security/users | :green_circle: | |
| POST /security/roles | :green_circle: | |
| POST /security/policies | :green_circle: | |
| POST /security/rules | :green_circle: | |
| PUT /agents/group | :red_circle: | #6749 (Fixed in `4.3.0`) |
| PUT /agents/group/new_test_group/restart | :red_circle: | #8870 (Fixed in `4.3.0`) |
| PUT /agents/restart | :red_circle: | #8870 (Fixed in `4.3.0`) |
| POST /agents | :green_circle: | |
| POST /agents/insert​|  :green_circle: | |
| POST /agents/insert​/quick | :green_circle: | |
| GET /agents | :green_circle: | |
| GET /agents/no_group | :green_circle: | |
| GET /agents/outdated | :green_circle: | |
| GET /decoders | :green_circle: | |
| GET /groups | :green_circle: | |
| GET /lists | :green_circle: | |
| GET /rules | :green_circle: | |
| GET /security/users | :green_circle: | |
| GET /security/roles | :green_circle: | |
| GET /security/policies | :green_circle: | |
| GET /security/rules | :green_circle: | |
| DELETE /agents | :red_circle: | #8853 (Fixed in `4.3.0`) |
| DELETE /groups | :red_circle: | #9014 (Fixed in `4.3.0`) |
| DELETE /security/users | :green_circle: | |
| DELETE /security/roles | :green_circle: | |
| DELETE /security/policies | :green_circle: | |
| DELETE /security/rules | :green_circle: | |
| PUT /manager/restart | :green_circle: | |
| PUT /cluster/restart | :green_circle: | |

## To be included in triage for rc13:
- [ ] https://github.com/wazuh/wazuh/issues/9451 (Healthcheck)
- [ ] https://github.com/wazuh/wazuh/issues/9442 (test_agent_PUT_endpoints.tavern.yaml)
- [x] https://github.com/wazuh/wazuh/issues/9458 (test_syscollector_endpoints.tavern.yaml)

**Comment by AdriiiPRodri** [comment]
## Status update

These are the results of the load tests on the cluster. We are going to check the CPU, memory, and disk usage (for more statistics see the attached zip file). In addition, we are going to check the cluster synchronization times for the agent-sync, integrity-sync, and integrity-check tasks.

Finally, we will analyze the results and provide a conclusion.

### 1000 agents and 2 worker nodes

#### CPU usage
![image](https://user-images.githubusercontent.com/15522808/128869244-05056ae4-f9ef-4bd6-81f3-d727a3988507.png)


#### Memory usage
![image](https://user-images.githubusercontent.com/15522808/128869296-45caf4b3-16c7-48c3-9c71-a54d69f1c62a.png)


#### Disk usage
![image](https://user-images.githubusercontent.com/15522808/128869333-81107bfa-2974-4465-b954-62a500478cbc.png)


#### Synchronization tasks
##### Agent-info
![image](https://user-images.githubusercontent.com/15522808/128869388-c02b361f-3a1e-4ae4-801b-ba792779872f.png)

##### Integrity-check
![image](https://user-images.githubusercontent.com/15522808/128869452-06baf5ae-5714-4138-8024-92078c38e714.png)

##### Integrity-sync
![image](https://user-images.githubusercontent.com/15522808/128869496-a59ef9f5-1bde-4aca-8a00-1c25654f1669.png)


### 10000 agents and 10 worker nodes

#### CPU usage
![image](https://user-images.githubusercontent.com/15522808/128869564-3ef4fa60-92bd-407d-b432-51cd9906ac12.png)


#### Memory usage
![image](https://user-images.githubusercontent.com/15522808/128869614-fb9f9c05-1191-41fb-b4fd-ea9f8af82454.png)


#### Disk usage
![image](https://user-images.githubusercontent.com/15522808/128869650-32ebb66f-83ad-472d-b01a-803e1faef8f1.png)


#### Synchronization tasks

##### Agent-info
![image](https://user-images.githubusercontent.com/15522808/128869732-52f017d8-5ef8-4860-96b3-378a8abf0ede.png)


##### Integrity-check
![image](https://user-images.githubusercontent.com/15522808/128869775-761638f6-0ac7-4d9f-8a3a-94ff5966ea2f.png)


##### Integrity-sync
![image](https://user-images.githubusercontent.com/15522808/128869809-0f723ede-7859-46a6-8221-1b94ddafa793.png)


### Conclusions

On the one hand, we will first analyze the charts obtained for CPU usage, percentage of disk used and RAM used. As for the CPU, we can see that there are usage peaks, especially in the API, these peaks correspond to the requests made to the API during the test, once answered we see how the CPU is released. We believe that these values are adequate.

As for the disk usage, we see that the cluster uses the disk during the synchronization of files, once the files are synchronized we can see how the disk usage drops. These increases occur during synchronization. Again, we consider these values normal and see nothing to indicate that there may be some kind of leak.

Finally, we have the RAM usage, to check this value we will look at the USS chart. In this chart, we see how the API usually consumes about **100MB** of memory, and in none of the tests, it increases more than this value. Regarding the cluster, we can see that the RAM memory used varies, this is due to the fact that we vary the number of agents and worker nodes. There is no constant increase that could indicate a memory leak.

On the other hand, we have cluster synchronization values. We will focus on the results of the tests with **10000** agents. In these, we see that the average synchronization time (integrity_sync) is **0.622** seconds, with a median of **0.516** seconds and a maximum peak of about **3** seconds, which shows the improvements made in `4.2`. As for the integrity_check task time, we have that the average time is **0.116** seconds and a median of **0.115**, this task has improved its performance significantly in the development of `4.2`, using an auxiliary variable to avoid unnecessary checks. Finally, we have the agent_sync task, which has also improved substantially, having a mean time of only `0.071` seconds and a median of **0.059**, having a peak of about **1** second.

All these values seem to be correct, being the highest the `integrity_sync task`, however, this task does not scale linearly depending on the agents, so the values are correct and we do not see any performance issue.

**The analysis with 50,000 agents is still pending.**

**Comment by AdriiiPRodri** [comment]
## Status update

Next, we will analyze the results for the 50000 agent tests.

### 50000 agents and 10 worker nodes

#### CPU usage
![image](https://user-images.githubusercontent.com/15522808/128978794-628ec09c-4764-4872-b7a8-4df926929b5c.png)


#### Memory usage
![image](https://user-images.githubusercontent.com/15522808/128978830-738755ca-b533-4088-8ec6-a6f0119d9458.png)


#### Disk usage
![image](https://user-images.githubusercontent.com/15522808/128978865-ba2cafbc-1335-4d03-aa33-f791c1b4a9f4.png)


#### Synchronization tasks
##### Agent-info
![image](https://user-images.githubusercontent.com/15522808/128978912-582c3d53-4b22-40d3-bcc4-168b356cca79.png)

##### Integrity-check
![image](https://user-images.githubusercontent.com/15522808/128979031-ea3a1ee1-5537-450d-b5ec-33ae68c5f7d1.png)

##### Integrity-sync
![image](https://user-images.githubusercontent.com/15522808/128979072-e92c2349-95a6-4a26-bc15-62a96961525d.png)

### 50000 agents and 25 worker nodes

#### CPU usage
![image](https://user-images.githubusercontent.com/15522808/128979111-0003de2c-cfcf-45d5-a3f5-22353e82b847.png)


#### Memory usage
![image](https://user-images.githubusercontent.com/15522808/128979131-25171a93-aa89-401c-84cf-d22e363046f4.png)


#### Disk usage
![image](https://user-images.githubusercontent.com/15522808/128979152-6bd20976-e68d-4972-8d12-b1335ec38275.png)


#### Synchronization tasks

##### Agent-info
![image](https://user-images.githubusercontent.com/15522808/128979188-4f3883e4-2db8-42a3-92c4-eee50f6dd27c.png)

##### Integrity-check
![image](https://user-images.githubusercontent.com/15522808/128979238-960567ef-0ef3-4ae9-ac4c-72e35bb1d548.png)

##### Integrity-sync
![image](https://user-images.githubusercontent.com/15522808/128979267-6c85a77a-71d4-4b68-bd61-112921efdca5.png)

We can see how the CPU usage has increased for the cluster having 50000 agents, this is because the synchronization processes are more complex because more files must be analyzed than in the previous tests. We can also see how as the synchronization processes are finishing, the CPU usage used by the cluster decreases drastically. Each peak in the graphs coincides with the synchronization tasks. We consider these values normal and see no indication of any leaks.

As for the RAM used, we see that the RAM used varies according to the number of workers, being 600MB maximum for 50000 agents and 10 workers and 1GB maximum for 50000 agents and 25 workers. This usage is partly due to the synchronization processes, and again we see how during the initial synchronization, until all the worker nodes have all their information, the RAM usage increases step by step to finally be released when the process is finished. In version 4.1, this same test would have a different result because, in that version, the RAM was not released completely so as time passed, the RAM usage of the cluster increased considerably, this is solved as we can see in 4.2.

Finally, we have disk usage. Both graphs are very similar and again we have that until the synchronization of all the agents is not finished, the disk usage increases in each synchronization process until finally it is no longer used because all the information is already synchronized. Again, we can verify that there is no leak.

Regarding the charts related to the synchronization processes, we have that in both environments, the charts of the **agent-info_sync** and **integrity_check** tasks are very similar in time, having higher peaks in the environment with 25 workers. We can see that the integrity_check task has taken about 60 seconds in one of the peaks to finally go down to consume between 0 and 1 seconds. Again, although it is quite long, everything returns to normal values once the initial process is finished. In our test, all the agents are added and connected simultaneously, so it is a very aggressive test, as in real environments this rarely occurs. In any case, once the initial synchronization is finished, we see how the times go down and remain stable, so we consider that everything is correct.

Finally, we have the **integrity_sync** chart. In this we see how the times vary depending on the number of workers, having a maximum peak of about 80 seconds in the test with 25 workers. This is due to the large number of files to synchronize as well as the saturation produced by sending them to the worker nodes. Again we see how after the process the values drop, indicating that everything works correctly and once the initial synchronization is done, the subsequent synchronizations are much faster.

### Reports
- 50000 agents 10 workers: 
[50000ag10wk.zip](https://github.com/wazuh/wazuh/files/6966283/50000ag10wk.zip)

- 50000 agents 25 workers: 
[50000ag25wk.zip](https://github.com/wazuh/wazuh/files/6966284/50000ag25wk.zip)
